proposals/0002-gep.md

```diff
@@ -0,0 +1,140 @@
+---
+MEP: 0002
+Title: Unifed API
+Discussion: Link
+Implementation: Link
+---
+
+# TITLE: Unified API
+
+## Abstract
+
+This GEP discusses the unified API that handles the interaction with LLM providers. This represents the feature set for MVP.
+
+## Motivation
+
+
+The Unified API abstracts much of the overhead of interacting with LLM providers.
+It stores all the provider specific logic in a single place and allowing developers to focus on the core logic. It also introduces durability to LLM apps. Additionally, unifying the API allows optimizations to be leveraged such as latency and cost improvements.
+
+
+### Requirements
+
+- R1: User should be able to specify specific models or select auto mode (selection handled by Glide)
+- R2: Should be a Hertz Client
+- R3: Should take a user request as input
+- R4: Handles all provider specific logic
+- R5: Easily maintained
+- R6: API schemas must unify common LLM request params (e.g. topP, N, etc)
+- R7: API routes must unify common LLM endpoints/API
+- R8: API schemas should allow to customize provider-specific params (e.g. a different messages for OpenAI/Anthropic)
+
+
+## Design
+
+
+- A Hertz client that handles an incoming request and returns a reponse from that provider.
+- Provider specific clients
+
+### Routes
+```yaml
+routes:
+    chat: /chat/{pool-id}
+    complete: /complete/{pool-id} #this is also know as generate in some providers
+    multi-modal: /multimodal/{pool-id} # future route, not for MVP
```

Comment:
Just want to double check: would a pool be able to handle all four types of API we want to provide? Other words, are all providers we support allows to do all four types of requests?

Reply:
No I don't think so at this point. All should support chat, complete but multimodal and embedding are likely not all support. I'm struggling to understand a good justification to include embedding. I'll keep thinking on it.

Reply:
> No I don't think so at this point. All should support chat, complete but multimodal and embedding are likely not all support.

That's good to know. Then at some point we will need to introduce pool types to match Glide API with a pool that support it e.g. multimodal API should be available for pools that strictly consist of models/providers that support this type of interactions

> I'm struggling to understand a good justification to include embedding

I have been thinking about this one from the resiliency perspective too. For example, if we have a pool of OpenAI & Anthropic embedders and we configure load balancing on it, would this be a problem that a part of docs are encoded via one embedder and another part is via another?

This still could be useful to decouple application from the underlying provider/model via Glide Unified API.

Reply:
Yeah that would be an issues. Need to embed docs with the same model. Here are my main reasons the deprioritize the embeddings.

1. Embedding for many apps currently are a one time process (I expect this to change long term)
2. You need to embed with the same model across all docs
3. Embedding APIs are less susceptible to issues seen with the LLM APIs
4. Chat/Complete is far more valuable to the end user

As an end user of Glide, I couldn't care less if embeddings were supported.

Reply:
@mkrueger12 sounds reasonable Max. Let's make just record this decision to postpone embeddings API until it's absolutely clear we need them.